### DISCUSSION TITLE
A horrible hack to optimize `U256` compile time

### DISCUSSION THREAD

**Discussion by Kixunil** [rust-bitcoin/rust-bitcoin#3021] [Ideas]
Firstly, this is mainly a crazy idea and I can't tell right now if it's worth it (probably not but it might be).

I've realized that after making C impl of `secp256k1` optional in some scenarios the whole `u256` code gets compiled twice - specifically if you depend on both `units` and `secp256k1` with the C library disabled. It also needs to be maintained twice. The ideal solution for this would be [fallback features](https://github.com/rust-lang/cargo/issues/1839) however that is not really "around the corner".

My idea is: expose `U256` from `secp256k1` behind a feature flag and require a different feature flag (e.g. `crypto-types`, on by default) to enable the rest of the library. Then `units` will depend on `secp256k1` to get `U256` (with the rest off) and keys will depend on `secp256k1` without `U256`. The `U256` code will use `#[cfg(any(feature = "u256", feature = "crypto-types"))]` to control compilation.

So as long as you have a single version of `secp256k1` you compile that code only once.

**Comment by apoelstra** [comment]
I think it's probably not worth it. Aside from the weirdly-shaped depenedency tree it creates (`units` really should not have crypto dependencies) I don't think there's actually a lot of reused code here, because the libsecp U256 does a bunch of modular arithmetic and the units U256 does a few target/difficulty-specific things.

**Reply by Kixunil** [reply]
Yeah, I suspect it compiles super-fast. We can still share the code in ten years when/if they add fallback features. The most annoying problem is having two copies in two repositories. I think we should figure out some CI action to diff them or, God forbid, git submodules.
